### PR TITLE
fix: fix temperature-related issues

### LINKED ISSUE #902: Fix and test accuracies related to temperature and vllm final returned logprob
As discussed in the comments of https://github.com/NVIDIA-NeMo/RL/pull/773, current `vllm==0.10.0` returns the logprobs without processing temperature, top-k, top-p, etc., which may lead to off-policy situations.

This should not affect training with `temperature=1.0`, default top-k and top-k.

Start with GRPO (AceReason, Deepscaler, Qwen 2.5 7B) https://github.com/NVIDIA-NeMo/RL/issues/887.

### PR DESCRIPTION
**Pull request by zhandaz**
# What does this PR do ?

Fix the temperature-related accuracy issues in both the dtensor path and the mcore path.

As reported in #887 and some other researchers, training GRPO (some variants of Qwen 2.5 7B, AceMath) does not work.

```bash
uv run python -u NeMo-Skills/nemo_skills/training/nemo_rl/start_grpo.py \
    ++policy.model_name=models/final_hf_checkpoint_1220 \
    ++cluster.gpus_per_node=8 \
    ++cluster.num_nodes=${NUM_ACTOR_NODES} \
    +data.train_data_path=data/deepseek_r1/rl_data/acemath_rl/shuffled_acemath_rl_920k.jsonl \
    +data.val_data_path=data/deepseek_r1/rl_data/acemath_rl/shuffled_acemath_rl_920k.jsonl \
    ++checkpointing.checkpoint_dir=results/${TASK_NAME}_${EXP_SUFFIX}/checkpoints \
    ++logger.log_dir=results/${TASK_NAME}_${EXP_SUFFIX}/training-logs \
    ++logger.wandb_enabled=True \
    ++logger.wandb.project=${WANDB_PROJECT} \
    ++logger.wandb.name=${WANDB_NAME} \
    ++logger.wandb.id=${WANDB_NAME} \
    ++policy.train_global_batch_size=1024 \
    ++policy.train_micro_batch_size=1 \
    ++policy.max_total_sequence_length=8192 \
    ++policy.logprob_batch_size=4 \
    ++policy.sequence_packing.enabled=True \
    ++policy.sequence_packing.train_mb_tokens=8192 \
    ++policy.sequence_packing.logprob_mb_tokens=8192 \
    ++policy.sequence_packing.algorithm="modified_first_fit_decreasing" \
    ++policy.sequence_packing.sequence_length_round=64 \
    ++policy.generation.vllm_cfg.tensor_parallel_size=4 \
    ++policy.generation.vllm_cfg.gpu_memory_utilization=0.8 \
    ++policy.generation.vllm_cfg.pipeline_parallel_size=1 \
    ++policy.generation.temperature=0.6 \
    ++policy.optimizer.kwargs.lr=1e-06 \
    ++loss_fn.reference_policy_kl_penalty=0.001 \
    ++policy.dtensor_cfg.enabled=True \
    ++policy.dtensor_cfg.tensor_parallel_size=2 \
    ++policy.dtensor_cfg.sequence_parallel=False \
    ++policy.dtensor_cfg.activation_checkpointing=True \
    ++data.prompt.prompt_config=qwen/math-cot \
    ++data.prompt.prompt_template=qwen-instruct \
    ++grpo.num_prompts_per_step=128 \
    ++grpo.num_generations_per_prompt=8
```

### 1. Reproduce the errors

As we can see in the images below and the [wandb](https://wandb.ai/nv-default-onboard/zhandaz-nemo-rl-dev/runs/zhanda-nemo-rl-7b-t-0.6-stage-1-1-baseline_exp_07), the rewards decrease a lot after around 50 steps, and the `token_mult_prob_error` ([meaning](https://github.com/NVIDIA-NeMo/RL/blob/70b966677a7fccaa0bce8b97c1df349e1dccaae5/docs/guides/grpo.md#multiplicative-token-probability-error-token_mult_prob_error)) is large.

Image

### 2. Fix the temperature error in the DTensor Policy Worker

We found the reason could be: in this PR https://github.com/NVIDIA-NeMo/RL/pull/660, it scales wrt the temperature according to the vllm engine version (as vllm V1 engine does not return the final logprobs wrt sampling parameters). This fix is actually wrong. We should follow the pattern in https://github.com/NVIDIA-NeMo/RL/pull/316 to scale logits according to the temperatures anyway. Otherwise, it may lead to off-policy. The inference and the training behavior do not match.

By removing the `if` in https://github.com/NVIDIA-NeMo/RL/blob/df31c1bf428255e982303e69adaa19058c2a7970/nemo_rl/models/policy/dtensor_policy_worker.py#L471-L472
we can see that the reward is back to normal, and `token_mult_prob_error` is much reasonable (though it still has one large spike that is larger than 2).

Image

### 3. Fix the returned logprobs in vLLM V1

We know that the returned logprobs from vLLM V1 is not correct for post-training because it is the raw logits (before applying any sampling paramters). Therefore, the `token_mult_prob_error` is actually an inaccurate indicator. **The convergence should not be affected since `use_importance_sampling_correction` and `use_on_policy_kl_approximation` are False, and the vllm returned logprobs are not involved in the real training (just used in the metrics calculation).** 

To confirm that, we do a file patch over the vLLM to return the logprobs after temperature. The reward curve remains almost the same in case 2, and the `token_mult_prob_error` is almost around 1 (which is expected).

Image

### Related

Top-p and Top-k fixing are not supported yet. It is not quite elegant to patch vllm to support top-p and top-k as it involves patching multiple files and many code blocks. We have more discussion in https://github.com/NVIDIA-NeMo/RL/pull/773 and https://github.com/NVIDIA-NeMo/RL/issues/69.

# Issues

Fixes #902, #887

### CODE REVIEW

File: nemo_rl/models/generation/vllm.py [hunk 1]
```diff
@@ -275,6 +275,43 @@ def _patch_vllm_init_workers_ray():
 
             _patch_vllm_init_workers_ray()
 
+            # Patch the vLLM sampler.py file to modify logprobs computation wrt temperature.
+            # This replaces raw_logprobs = self.compute_logprobs(logits) with custom temperature-applied logprobs.
+            # TODO(zhanda): This is only a temporary fix to address the issue of incorrect logprobs returned by vllm
+            # and should be removed or improved after vllm's new logprobs option is released. And currently, other
+            # sampling parameters like top_p, top_k, etc. are not supported.
```
Comment (parthchadha) [inline nemo_rl/models/generation/vllm.py]: @Dazz993 do you mind adding assert errors regarding `top_p` and `top_k` not supported?

File: nemo_rl/models/policy/dtensor_policy_worker.py [hunk 2]
```diff
@@ -463,13 +462,8 @@ def create_context_parallel_ctx(
...
 
     def _apply_temperature_scaling(self, logits: torch.Tensor) -> torch.Tensor:
-        # Apply temperature scaling to logits if configured and not using V1 engine.
         if "generation" in self.cfg and self.cfg["generation"] is not None:
-            # The V1 engine returns raw logits before temperature scaling.
-            # The V0 engine returns scaled logits.
-            # Therefore, we only divide if we are not using the V1 engine.
-            if not is_vllm_v1_engine_enabled():
-                logits.div_(self.cfg["generation"]["temperature"])
+            logits.div_(self.cfg["generation"]["temperature"])
```
Comment (parthchadha) [inline nemo_rl/models/policy/dtensor_policy_worker.py:478]: @Dazz993 looks like we are not handling temperature in mcore worker (https://github.com/NVIDIA-NeMo/RL/blob/70b966677a7fccaa0bce8b97c1df349e1dccaae5/nemo_rl/models/policy/megatron_policy_worker.py#L132). Not blocking this PR but can add this fix/rerun experiments with mcore worker?

**Review by parthchadha — DISMISSED** [review]

File: nemo_rl/models/generation/vllm.py (hunk 1 above)
Comment (zhandaz) [reply nemo_rl/models/generation/vllm.py]: Added.

File: nemo_rl/models/policy/dtensor_policy_worker.py (hunk 2 above)
Comment (zhandaz) [reply nemo_rl/models/policy/dtensor_policy_worker.py:478]: Sure. Will do in another PR since I have to rerun the experiments and may take some time.

**Review by parthchadha — DISMISSED** [review]

**Comment by zhandaz** [comment]
Have rerun the experiments after merging the main. There is a small spike. Should be fine.
image

**Review by wangshangsam — APPROVED** [review]
LGTM!

I don't have further comments but I just have a quick question - does it mean that, once vLLM 0.11 is released, we don't need to patch
`raw_logprobs = self.compute_logprobs(logits)` into 
`raw_logprobs = self.compute_logprobs(self.apply_temperature(logits.to(torch.float32), sampling_metadata.temperature))` entirely?

**Comment by zhandaz** [comment]
@wangshangsam @parthchadha I found the tests fail in this line: 
https://github.com/NVIDIA-NeMo/RL/blob/33ca684e123b6bb24e83927b95fa048a0fed79b6/tests/unit/models/generation/test_vllm_generation.py#L722

And the reason for that is in the test, we generate results with `greedy=True`:
https://github.com/NVIDIA-NeMo/RL/blob/33ca684e123b6bb24e83927b95fa048a0fed79b6/tests/unit/models/generation/test_vllm_generation.py#L659-L666

which makes vllm to set `temperature` to be None. However, when we do the training or get_logprobs from the training policy, we don't have `greedy=True`. Therefore, the logprobs would not match. As far as I know, `greedy=True` is only used in the tests.

To solve this problem, we can:
- Option 1: Simply remove the `greedy=True` in this `run_hf_train_process` test function.
- Option 2: add the `greedy` argument to related function in `lm_policy` as well.

**Comment by zhandaz** [comment]
> It seems to me that the purpose of `greedy=True` is to make the generation deterministic (i.e., so that the results from HF and vLLM would match)? Therefore, I'm not sure if option 1 is gonna work.

As we discussed offline, whether `greedy` is True or False does not necessarily ensure the results from HF and vLLM match. @parthchadha what do you think?

**Review by parthchadha — APPROVED** [review]
